### PR TITLE
tests: parse timestamps before comparing them

### LINKED ISSUE #4003: Race condition in unit tests using cmpopts.EquateApproxTime
### What happened?

I ran `make test` and curiously got an error without even modifying the code. I looked deeper into it and it looks like tests that use `cmpopts.EquateApproxTime` will fail (sometimes) if executed repeatedly.
The two affected packages are `internal/controller/apiextensions/composite` and `internal/controller/apiextensions/claim`.

### How can we reproduce it?

```
go test -count=1000 ./internal/controller/apiextensions/composite
go test -count=1000 ./internal/controller/apiextensions/claim
```

<details><summary>Test failure from ./internal/controller/apiextensions/composite</summary>

```nocode
--- FAIL: TestReconcile (0.00s)
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "spec":       map[string]any{"compositionRef": map[string]any{}},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "message":            string("cannot compose resources: boom"),
                                                        "reason":             string("ReconcileError"),
                                                        ... // 2 identical entries
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "spec":       map[string]any{"compositionRef": map[string]any{}},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "message":            string("cannot publish connection details: boom"),
                                                        "reason":             string("ReconcileError"),
                                                        ... // 2 identical entries
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "spec":       map[string]any{"compositionRef": map[string]any{}},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("ReconcileSuccess"),
                                                        "status":             string("True"),
                                                        "type":               string("Synced"),
                                                },
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("Available"),
                                                        "status":             string("True"),
                                                        "type":               string("Ready"),
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "spec":       map[string]any{"compositionRef": map[string]any{}},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "message":            string("cannot fetch Composition: boom"),
                                                        "reason":             string("ReconcileError"),
                                                        ... // 2 identical entries
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "spec":       map[string]any{"compositionRef": map[string]any{}},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "message":            string("cannot configure composite resource: boom"),
                                                        "reason":             string("ReconcileError"),
                                                        ... // 2 identical entries
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                ... // 2 identical entries
                                "metadata": map[string]any{"annotations": map[string]any{"crossplane.io/paused": string("")}},
                                "spec":     map[string]any{"compositionRef": map[string]any{}},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("ReconcileSuccess"),
                                                        "status":             string("True"),
                                                        "type":               string("Synced"),
                                                },
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("Available"),
                                                        "status":             string("True"),
                                                        "type":               string("Ready"),
                                                },
                                        },
        -                               "connectionDetails": map[string]any{"lastPublishedTime": string("2023-04-21T09:46:26Z")},
        +                               "connectionDetails": map[string]any{"lastPublishedTime": string("2023-04-21T09:46:27Z")},
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "metadata":   map[string]any{"annotations": map[string]any{"crossplane.io/paused": string("true")}},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("ReconcilePaused"),
                                                        "status":             string("False"),
                                                        "type":               string("Synced"),
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "metadata":   map[string]any{"deletionTimestamp": string("2023-04-21T09:46:26Z")},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("Deleting"),
                                                        "status":             string("False"),
                                                        "type":               string("Ready"),
                                                },
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "message":            string("cannot unpublish connection details: boom"),
                                                        "reason":             string("ReconcileError"),
                                                        ... // 2 identical entries
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "metadata":   map[string]any{"deletionTimestamp": string("2023-04-21T09:46:26Z")},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("Deleting"),
                                                        "status":             string("False"),
                                                        "type":               string("Ready"),
                                                },
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("ReconcileSuccess"),
                                                        "status":             string("True"),
                                                        "type":               string("Synced"),
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "spec":       map[string]any{"compositionRef": map[string]any{}},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "message":            string("refusing to use invalid Composition: boom"),
                                                        "reason":             string("ReconcileError"),
                                                        ... // 2 identical entries
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "message":            string("cannot select CompositionUpdatePolicy: boom"),
                                                        "reason":             string("ReconcileError"),
                                                        ... // 2 identical entries
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "spec":       map[string]any{"compositionRef": map[string]any{}},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("ReconcileSuccess"),
                                                        "status":             string("True"),
                                                        "type":               string("Synced"),
                                                },
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("Creating"),
                                                        "status":             string("False"),
                                                        "type":               string("Ready"),
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "message":            string("cannot select Composition: boom"),
                                                        "reason":             string("ReconcileError"),
                                                        ... // 2 identical entries
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "spec":       map[string]any{"compositionRef": map[string]any{}},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("ReconcileSuccess"),
                                                        "status":             string("True"),
                                                        "type":               string("Synced"),
                                                },
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("Available"),
                                                        "status":             string("True"),
                                                        "type":               string("Ready"),
                                                },
                                        },
        -                               "connectionDetails": map[string]any{"lastPublishedTime": string("2023-04-21T09:46:26Z")},
        +                               "connectionDetails": map[string]any{"lastPublishedTime": string("2023-04-21T09:46:27Z")},
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "spec":       map[string]any{"compositionRef": map[string]any{}},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("ReconcileSuccess"),
                                                        "status":             string("True"),
                                                        "type":               string("Synced"),
                                                },
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("Available"),
                                                        "status":             string("True"),
                                                        "type":               string("Ready"),
                                                },
                                        },
        -                               "connectionDetails": map[string]any{"lastPublishedTime": string("2023-04-21T09:46:26Z")},
        +                               "connectionDetails": map[string]any{"lastPublishedTime": string("2023-04-21T09:46:27Z")},
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "metadata":   map[string]any{"deletionTimestamp": string("2023-04-21T09:46:26Z")},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("Deleting"),
                                                        "status":             string("False"),
                                                        "type":               string("Ready"),
                                                },
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "message":            string("cannot remove composite resource finalizer: boom"),
                                                        "reason":             string("ReconcileError"),
                                                        ... // 2 identical entries
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "message":            string("cannot add composite resource finalizer: boom"),
                                                        "reason":             string("ReconcileError"),
                                                        ... // 2 identical entries
                                                },
                                        },
                                },
                        },
                },
          }
FAIL
FAIL    github.com/crossplane/crossplane/internal/controller/apiextensions/composite    0.035s
FAIL
```

</details>

### PR DESCRIPTION
**Pull request by phisco**
### Description of your changes

Fixes #4003.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

Unit tests are passing and checked with the debugger that EquateApproxTime is actually executed now, which was not the case before.

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: internal/controller/apiextensions/claim/reconciler_test.go [hunk 1]
```diff
@@ -825,7 +825,13 @@ func TestReconcile(t *testing.T) {
 
 			got, err := r.Reconcile(context.Background(), reconcile.Request{})
 
-			if diff := cmp.Diff(tc.want.claim, tc.args.claim, cmpopts.EquateApproxTime(30*time.Second)); diff != "" {
+			if diff := cmp.Diff(tc.want.claim, tc.args.claim, cmpopts.AcyclicTransformer("StringToTime", func(s string) any {
```
Comment (hasheddan) [inline internal/controller/apiextensions/claim/reconciler_test.go:828]: Might be useful to add this transformer to `crossplane-runtime` in the future!

File: internal/controller/apiextensions/composite/reconciler_test.go [hunk 2]
```diff
@@ -775,7 +775,13 @@ func WantComposite(t *testing.T, want resource.Composite) func(ctx context.Conte
...
 		// lastTransitionTime, but we may be using unstructured types here where
 		// the conditions are just a map[string]any.
-		diff := cmp.Diff(want, got, cmpopts.EquateApproxTime(3*time.Second))
+		diff := cmp.Diff(want, got, cmpopts.AcyclicTransformer("StringToTime", func(s string) any {
+			ts, err := time.Parse(time.RFC3339, s)
+			if err != nil {
+				return s
+			}
+			return ts
+		}), cmpopts.EquateApproxTime(3*time.Second))
```
Comment (hasheddan) [inline internal/controller/apiextensions/composite/reconciler_test.go:784]: Is the approx time here intentionally less than the other occurrence? (30 seconds vs 3 seconds)

Comment (phisco) [reply internal/controller/apiextensions/composite/reconciler_test.go:784]: Not sure about that tbh, I don't think there was a particular reason, and then we had a 99.9% success with 0s approximation so far, so I guess 3s could be enough 🤔

File: internal/controller/apiextensions/claim/reconciler_test.go (hunk 1 above)
Comment (phisco) [reply internal/controller/apiextensions/claim/reconciler_test.go:828]: 👍 if we agree on a final form here, I can freeze this PR, add it to crossplane-runtime, release a new version and switch this to consume that instead.

Comment (hasheddan) [reply internal/controller/apiextensions/claim/reconciler_test.go:828]: @phisco I don't think this needs to be blocker for fixing up these tests 👍🏻

File: internal/controller/apiextensions/composite/reconciler_test.go (hunk 2 above)
Comment (hasheddan) [reply internal/controller/apiextensions/composite/reconciler_test.go:784]: @phisco makes sense -- I know @negz mentioned on the PR where I added the approx time that `30s` felt like a lot anyway. Can you update to standardize on `3s` in all locations?

Comment (phisco) [reply internal/controller/apiextensions/composite/reconciler_test.go:784]: Done!

**Review by hasheddan — APPROVED** [review]

**Pull request merged**